src/script/sigcache.cpp

```diff
@@ -109,17 +126,71 @@ void InitSignatureCache()
         (nElems * sizeof(uint256)) >> 20, nMaxCacheSize >> 20, nElems);
 }
 
+template <typename F>
+bool RunMemoizedCheck(const std::vector<uint8_t> &vchSig,
+    const CPubKey &pubkey,
+    const uint256 &sighash,
+    uint32_t flags,
+    bool storeOrErase,
+    const F &fun)
+{
+    uint256 entry;
+    signatureCache.ComputeEntry(entry, vchSig, pubkey, sighash, flags);
+    if (signatureCache.Get(entry, !storeOrErase))
+    {
+        return true;
+    }
+
+    if (!fun())
```

Comment:
is the code above the call to this passed function so expensive?
or is the code of the passed function expensive that you rather defer it?, say:
`return TransactionSignatureChecker::VerifySignature(vchSig, pubkey, sighash); ` is one that I see you pass.

if not, perhaps it's simpler to just pass the bool output of the function as a parameter, even if it's not to be used.
It's a pretty cool optimization, but when I see that sometimes you pass a lambda that returns `false` it's kind of weird.

Reply:
see my comment on the #ifdefed out below